---
title: "Short INDELs: genetic markers for adaptive divergence"
date: "5th March 2020"
author: "Samuel Perini et al."
---

## Introduction

Since the 1960s, researchers have used genetic variants as markers to study evolutionary processes. The first ones were protein variants in enzymes (allozymes) and were initially discovered in populations of humans and Drosophila using gel electrophoresis [@harris1966; @johnson1966; @hubbyandlewontin1966]. Their application became broad in terms of number of species that could be analyzed but at the same time limited to measure indirectly DNA variation only in protein coding sequences. New DNA-based markers were developed that could capture directly DNA variation both inside and outside protein coding sequences [e.g., restriction fragment length polymorphisms @botsteinetal1980] and that could also be amplified via PCR [@saiki-et-al1985]. In population genetics studies, the most popular PCR-based markers were microsatellites [@jarneandlagoda1996; @goldstein-and-schlotterer1999]. However, after the emergence of high-throughput and next-generation sequencing technologies, genetic markers such as single nucleotide polymorphisms (SNPs), outcompeted microsatellites because of their lower mutation rate, higher abundance, broader genome coverage, easier bioinformatic automation and higher opportunity for comparative studies across taxa [@schlotterer2004]. SNPs, in particular, have been extensively used for studying the evolutionary history and adaptation of populations [@brumfield2003; @morin2004].  
Many current studies use SNPs to identify regions under divergent selection and have found that, compared to the genomic background, these regions showed high divergence between species (high F<sub>ST</sub>), low nucleotide diversity within species (low $\pi$), an excess of rare alleles (negative Tajimaâ€™s D) and strong levels of linkage disequilibrium (high _r_^2). Similar patterns can also be generated by background selection where, in areas of low recombination, SNP polymorphisms are reduced compared to other parts of the genome because of the removal of neutral variation linked to deleterious mutations [@charlesworth-et-al1993]. Furthermore, it is not guaranteed that in the presence of divergent natural selection or background selection, two different species/populations showed high F_ST. There have been empirical cases where genomic regions contributing to divergence were showed to contain clusters of SNPs in high linkage disequilibrium but with relatively low levels of differentiation [e.g., @westram2018]. Clearly, genomic patterns related to population divergence and speciation can be successfully captured using SNP markers but it is also evident that these are not the only variants present at significant frequencies in many populations [REF]. There are other elements of variation across the genome that can also be used as genetic markers for evolutionary processes. Short insertions and deletions (INDELs $\leq$ 50 bp) are not to be denied from having such a function as they are the most common type of genetic variability among the non-SNP variants [e.g., @montgomery2013].  
There might be limitations with the use of INDEL markers. Compared to point mutations, larger modifications to an already functional system are more likely to disrupt the function of a protein and cause a fitness damage. As a consequence, they will be maintained at low frequencies by the action of purifying selection [@fayetal2001; @massouras2012]. In line with these expectations, working with INDELs has the disadvantage of using a genetic marker that is twice less common than SNPs due to a higher probability of negative fitness effects [@clark2007]. Such a low frequency may not provide any additional information than just a signature of the action of purifying natural selection on INDELs. However, there are examples of INDELs that are maintained by either positive selection, as for human-specific INDELs with a putative role in promoting RNA and protein level changes [@chen2007], or balancing selection, as for the INDEL polymorphism for flowering time in _Arabidopsis lyrate_ [@kuittinen-et-al2008].  
Evolutionary geneticists have already recognized the role of INDELs in respect to ... [@chen2009; @bartonandzeng2019] but our knowledge is still limited to a few model species. Moreover, common practice does not entail the use of both SNPs and short INDELs when studying population history and adaptive divergence. Typical evolutionary genetic studies discard all variants apart from diallelic SNPs even though SNPs and INDELs can be detected simultaneously by the same variant calling algorithm. CITE A FEW RECENT EXAMPLES WHERE INDELS HAVE BEEN REMOVED.  
INDEL discovery does not come without bioinformatic challenges [@vali2008; @onishi2011] which are probably part of the reason why variant calling pipelines often adopt an INDEL filter. Mapping algorithms deal poorly with long INDELs and with repeated motifs [@narzisiandschatz2015]. False-negatives and false-positives can be generated when coverage distribution is not uniform or the efficiency of targeted resequencing is not even across the queried regions of the genome [@fang2016]. However, these issues are encountered less frequently with short INDELs as they showed a lower false discovery rate than longer ones [@massouras2012]. Using the current best practice for variant calling, short INDELs can be called relatively reliably and be easily integrated with SNPs [REF FOR GATK best practice; @hwangetal2015; @li2018; AND MORE].  
Hence, in order to understand whether studies of adaptive divergence can benefit from complementing SNPs with short INDELs, the first step is to assess whether the patterns of variation in short INDELs are coherent with the ones in SNPs. Here, we study a system of adaptive divergence and test whether using short INDELs as markers allows us to draw the same conclusions about biological processes as SNPs. We decided to study typical patterns of adaptive divergence in the intertidal snail _Littorina saxatilis_ because in this system there are replicated evidences for a heterogenous pattern of differentiation that is excellent for assessing the effects of population genomic processes on short INDELs, a type of variants that has been overlooked in this species.  
Genetic and genomic studies on _L. saxatilis_ have used a wide array of approaches, the latest being capture resequencing combined with hybrid zone analysis based on SNP markers [@westram2018, submitted]. This type of integration provides a powerful set of tools to investigate different forms of selection, their interaction and their relation with environmental factors and demography [@harrison1993]. For instance, when two populations live on distinct habitats but they still exchange genes at the contact zone, divergent natural selection and gene flow will have opposing effects on how allele frequencies change across the ecological gradient. When selection is stronger than gene flow, the change in allele frequencies will show a steeper cline than when selection is weaker. Therefore, measuring the slope of the allele frequency change (i.e., the slope at the cline center) can give us an estimate of the strength of selection which can then be used to identify divergently selected loci [@slatkin1973; @bartonandhewitt1985; @bartonandgale1993]. However, this also requires a substantial knowledge about the biology of the species. It is well documented that two populations of _L. saxatilis_, the 'Crab' and the 'Wave' ecotype, display divergent phenotypes that are associated with the ecological conditions of their local habitat. Snails with large shells and wary behavior are typically found in boulder fields where crab predation is common while snails with smaller shells and bolder behavior are instead mostly found in crevices of the bedrock where wave exposure is reduced. Such adaptations were shown experimentally to respond to crab predation and wave exposure, respectively [@johannesson1986; @lepennecetal2017].
Here, we specifically refer to a recent study (Westram et al. submitted) using SNP markers to study neutral and adaptive processes. This study compared seven replicate hybrid zones between _L. saxatilis_ ecotypes on a small geographical scale. SNPs were used to define the neutral population genetic structure. In addition, for each hybrid zone and SNP, cline analysis was performed to identify SNPs with steep allele frequency clines and / or large allele frequency differences between ecotypes ('outlier SNPs'). Such SNPs represent good candidates for genomic regions under divergent selection. The main patterns detected in this study were 1. extensive sharing of outliers between different hybrid zones, indicating a similar genomic basis of adaptation; 2. an accumulation of outliers in chromosomal inversions, suggesting that inversions contribute to local adaptation; 3. a shift of allele frequency clines away from the habitat transition into the Wave habitat. Here, we use this study as a case study to test whether the main findings can be replicated when using short INDELs as genetic markers. Given that short INDELs are expected to be under stronger purifying selection than SNPs, we also compare the proportions of SNPs and short INDELs that are putatively neutral, deleterious and beneficial. First, we annotate short variants across the _L. saxatilis_ genome and then, for each functional region, we infer the strength of natural selection using the allele frequency spectra.

## Materials and methods

SNPs are well-established genetic markers for population divergence [REF] which means that we can use their patterns of variation as a reference for comparison with the patterns of variation in short INDELs. Estimating to what extent patterns between SNPs and short INDELs differ can inform us about the plausibility of using short INDELs as supplementary or even alternative genetic markers of SNPs in studies of population genetics. Here, we analyzed a total of seven patterns of genetic variation using six hybrid zones. We first compared the proportion of SNPs and short INDELs that showed a clinal change in allele frequency, contrasted the frequencies of these clinal SNPs and clinal short INDELs at the cline ends, determined the positions of their cline centers and located these variants across the genome. We then measured the overall level of differentiation in SNPs and in short INDELs and compared their distribution of derived allele frequencies. Finally, we estimated to what extent the observed patterns in SNPs and the observed patterns in short INDELs were shared among the six hybrid zones.  
For SNPs, all the patterns of variation except for the distribution of derived allele frequencies, which was specific for this study, were obtained from Westram et al. (submitted). For short INDELs, the patterns were documented for the first time in this study. The raw data were obtained from Westram et al. (submitted) and the same analytical steps were performed whenever it was possible. Obviously, the bioinformatic pipeline had to be adjusted to include INDELs because Westram et al. (submitted) did not keep this type of genetic marker after variant calling. Any modifications to their bioinformatic pipeline was implemented to increase the accuracy of short INDEL discovery [@hwangetal2015] and these steps are described in the section Analysis of short genomic variants (see below). However, for those procedures where nothing was changed, we refer to Westram et al. (submitted) for details as they have been thoroughly described.  
Snails were sampled from the rocky shore of three different islands which were located on the Swedish west coast and labelled, for simplicity, CZA, CZB and CZD (Westram et al. submitted, Fig. 1A but note that ANG was not considered for the present study). Each shore contained two bedrocks separated by one boulder field, for a total of two Crab-Wave hybrid zones per island.  Per island, we sampled ~384 snails along the shore. In all following analyses, we consider the two hybrid zones on an island separately; i.e. we split the sampled transect in the middle of the boulder field to represent a 'left' and a 'right' hybrid zone. For each of the six hybrid zones, we defined the position of each snail on a one-dimensional path along the shore, making one-dimensional cline analysis possible. We also defined the location of the habitat transition from the boulder field to the bedrock on this path. DNA was extracted from a piece of foot tissue. A capture sequencing protocol was used to sequence 40,000 probe regions of a few hundred base pair length, randomly distributed across the _L. saxatilis_ genome. Further details can be found in Westram et al. (submitted).

_Analysis of short genomic variants_

The very first difference between this study and the work by Westram et al. (submitted) was that raw capture and paired-end sequencing reads from these authors were analyzed together with whole-genome sequences from three individuals of _L. compressa_, the chosen outgroup for _L. saxatilis_. These additional reads were needed in order to define the ancestral vs. derived state of alleles (see below, section _Shared consequences on fitness_).  
Trimming, mapping and quality filters stayed the same but variant calling was different. Westram et al. (submitted) called SNPs using SAMtools mpileup and BCFtools calls. However, because SAMtools [@lietal2009] has appeared to be less accurate than GATK [@mckennaetal2010] when it comes to INDEL discovery [@hwangetal2015], we decided to detect short variants (SNPs and INDELs $\leq$ 50 bp) following the GATK (version 4.1) best practice "Germline short variant discovery" (see https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-; last accessed February 4, 2020). This part of the pipeline consisted of three steps. In the first one, variants were called per-sample in local haplotypes which were assembled independently from the existing mapping information. In the second step, per-sample variants were combined and, in the third step, these were jointly genotyped [@vanderauweraetal2013; @poplinetal2017].  
The final adjustment that we made was to filter raw variants according to the GATK best practice hard filters (for SNPs: QD $<$ 2.0, QUAL $<$ 30.0, SOR $>$ 3.0, FS $>$ 60.0, MQ $<$ 40.0, MQRankSum $<$ -12.5, ReadPosRankSum $<$ -8.0; for INDELs: QD $<$ 2.0, QUAL $<$ 30.0, FS $>$ 200.0, ReadPosRankSum $<$ -20.0; see https://gatk.broadinstitute.org/hc/en-us/articles/360035531112#2).  
Filtered variants were then treated in the same way as the SNPs in Westram et al. (submitted). We retained only diallelic variants to reduce the number of false positive calls due to sequencing errors and we kept only the variants that were detected within 1,000 base pairs from any SNP present in the linkage map. By doing so, we excluded duplicated loci with unusual segregation patterns (see Westram et al. (submitted), Materials & Methods, section Genotyping and bioinformatics).
Other analytical and filtering steps that were in common with Westram et al. (submitted) were the ones involved in cline fitting, outlier identification and outlier location across the genome. Since we did not change any code from Westram et al. (submitted), we refer to the respective sections of their study for the details on how clines and outliers were estimated for both SNPs and short INDELs. The possibility of applying the same methods used for SNPs to short INDELs without any major adjustments means that future population genomic studies could easily investigate evolutionary processes using two of the major sources of genomic variation.

_Sharing of cline patterns_

Divergent natural selection and gene flow have opposing effects on how allele frequencies change across an ecological gradient. When selection is stronger than gene flow, the change in allele frequencies will show a steeper cline than when selection is weaker. Therefore, measuring the slope of the allele frequency change (i.e., the slope at the cline center) can give us an estimate of the strength of selection which can then be used to identify divergently selected loci (Slatkin 1973; Barton and Hewitt 1985; Barton and Gale 1993). However, Westram et al. (submitted) did not exclusively estimate the slope of the cline, or more precisely, the inverse of the cline width, but also the position of the cline center, the allele frequency at each cline end, and the variance in the data explained by the cline model. These five parameters were extracted after fitting a cline model for each individual SNP. For each individual short INDEL, we repeated the same analytical steps as in Westram et al. (submitted), Materials & Methods, section Cline analyses.  
To test whether the pattern of allele frequencies at cline ends and the positions of the cline centers were overlapping between clinal SNPs and clinal short INDELs, we computed, for each pattern, the joint frequency distribution between these variants.

_Sharing of outliers_

When a SNP marker falls in physical proximity to a locus that is under divergent natural selection, its pattern of segregation will resemble the one of the actual target of selection. This means that if we are able to identify these SNPs, we could regard them as if they were themselves the targets of selection and refer to them as the loci contributing to divergence. We could be even more confident that these SNPs or outliers were linked to selected loci if these were shared between independent locations that experienced similar selective pressures. For short INDELs, the same rationale should apply.  
In order to identify outlier SNPs, Westram et al. (submitted) first fitted clines for individual SNPs and then set a threshold that was based on a measure for the expected genetic variation that was explained by a model of neutral evolution. SNPs with a measure above this threshold were considered to have evolved non-neutrally and therefore have undergone the action of either direct selection or selection at linked loci. The authors have also calculated the proportion of non-neutral SNPs that were shared within and between islands. By performing these same analyses as described by Westram et al. (submitted) in Materials & Methods, section Outlier identification and sharing, we aimed to both identify a set of outliers for short INDELs and assess to what extent these were shared within and between islands.  
If we expect short INDELs to be linked to the targets of selection like SNPs, then we should be able to find outlier SNPs and outlier short INDELs to be shared in similar proportions within and between islands. We performed two-proportions z-test in R (R Core Team 2019) to compare proportions of outlier sharing between SNPs and short INDELs.

_Shared genomic architectures of divergence_

Regions of the genome that are characterized by reduced recombination rate have been shown, both theoretically and empirically, to facilitate population divergence despite gene flow [REF]. In agreement with these results, Westram et al. (submitted) found that outlier SNPs were overrepresented in regions of low recombination (i.e., inversions and centromeres) and that outliers in inversion regions were also shared the most between hybrid zones. We looked for the same patterns in outlier short INDELs and measured first the correlations between positions of outlier SNPs and outlier short INDELs on the linkage map. Secondly, we assessed whether outlier short INDELs that were shared among hybrid zones were predominant in inversions and other regions of low recombination.

_Shared consequences on fitness_

Another method that can be used for assessing to what degree SNPs and short INDELs are influenced by the same population genomic processes is to determine what consequences these variants have on fitness. For SNPs and short INDELs, we expected differences with respect to the strength of negative selection, especially in protein-coding regions, but we did not exclude the fact that, like for SNPs, most short INDELs will show signatures of neutral evolution and a few will show patterns of positive selection.  
We first annotated all SNPs and short INDELs with significant clines onto the _L. saxatilis_  GFF annotation file using SnpEff [@cingolanietal2012]. This tool classified each variant with a standardized term for a biological process such as "intron_variant" if a variant hit an intron, "missense_variant" if a variant changed one amino acid without affecting the length of the sequence and "frameshift_variant" if a variant disrupted the translational reading frame. We then grouped together all the annotated variants (SNPs + short INDELs) with either the same biological process (e.g., intron or missense variant) or the same effect at the amino acid level (e.g., nonsynonymous or frameshift). Finally, for each group, we computed the joint unfolded site-frequency spectrum between SNPs and short INDELs. Variants were polarized knowing their ancestral state in _L. compressa_ and it was executed using custom R script.

## Results


## Discussion


# References  
