#!/usr/bin/env python3

import os, glob


WDIR = "/proj/data9/samuel/2016_CZA_capture/Littorina_saxatilis/short_genetic_variants"

workdir: "/proj/data9/samuel/2016_CZA_capture/Littorina_saxatilis/short_genetic_variants"

FQ_DIR = WDIR + "/raw"
MQC_DIR = WDIR + "/mqc"
RES_DIR = WDIR + "/results"


WC = glob_wildcards(os.path.join(FQ_DIR, "{sample}_{pair}.fastq.gz"))
SAMPLES = set(WC.sample)
PAIR1, PAIR2 = set(WC.pair)


ALB_TRIM_DIR = "/usr/local/packages/Trimmomatic-0.36/trimmomatic-0.36.jar"

ALB_BED_DIR = "/usr/local/packages/anaconda2-2.5.0/bin/bedtools"

ALB_BWA_DIR = "/usr/local/packages/anaconda2-4.4.0/bin/bwa"

ALB_SAMT_DIR = "/usr/local/packages/anaconda2-5.0.0/bin/samtools"

ALB_SAMB_DIR = "/proj/data9/samuel/modules/sambamba_v0.6.6"

ALB_PIC_DIR = "/proj/data9/samuel/modules/picard/build/libs/picard.jar"

ALB_STA_DIR = "/proj/data9/samuel/modules/stampy-1.0.32/stampy.py"

ALB_STK_DIR = "/proj/data9/samuel/modules/seqtk/seqtk"

ALB_GATK_DIR = "/proj/data9/samuel/modules/gatk-4.0.2.0/gatk-package-4.0.2.0-local.jar"

ALB_CLIP_DIR = "/proj/data9/samuel/modules/bamUtil/bin/bam"



ADAPT = "/proj/data9/samuel/2016_CZA_capture/adapters/adapt_cont.fa:2:30:10"
WINDOW = "10:20"

REF = "/proj/data9/samuel/2016_CZA_capture/reference/Littorina_scaffolded_PacBio_run2_7_Oct_2016_unmasked.fasta"
RED_REF = "/proj/data9/samuel/2016_CZA_capture/red_ref/NEW_superscaffold_REF.fasta"



rule all:
	input:
		"genome/tmp_target.fasta",
		"genome/NONtarget.fasta",
		"red_ref/NEW_superscaffold_REF.fasta",
		"Target_contigs.bed",
		expand("results/gVCF/{S}.g.vcf.gz", S=SAMPLES),
		"results/VCF/CZA_raw_all.vcf.gz"




rule ref_prep:
	input: rules.cat_fasta.output
	output:
		"/proj/data9/samuel/2016_CZA_capture/reference/Littorina_scaffolded_PacBio_run2_7_Oct_2016_unmasked.fasta.dict"
	message:
		"""--- Preparation {input} with BWA index, samtools faidx, and Picard dict."""
	priority: 90
	threads: 10
	shell:
		"""
		/bin/sh scripts/ref_prep.sh {input} {output}
		"""

		
rule trimmomatic:
	input:
		fwd=expand("{fq_dir}/{{S}}_{pair}.fastq.gz", fq_dir=FQ_DIR, pair=PAIR1), 
		rev=expand("{fq_dir}/{{S}}_{pair}.fastq.gz", fq_dir=FQ_DIR, pair=PAIR2)
	output:
		fwd_pa=temp(expand("trimmed/{{S}}_{pair}_paired.fastq.gz", pair=PAIR1)),
		fwd_un=temp(expand("trimmed/{{S}}_{pair}_unpaired.fastq.gz", pair=PAIR1)),
		rev_pa=temp(expand("trimmed/{{S}}_{pair}_paired.fastq.gz", pair=PAIR2)),
		rev_un=temp(expand("trimmed/{{S}}_{pair}_unpaired.fastq.gz", pair=PAIR2))
	message: """--- Trimming Illumina adapters and PhiX from {input.fwd} and {input.rev} with Trimmomatic-0.36."""
	log:
		"logs/trimmomatic/{S}.log"
	params:
		leading=3,
		trailing=3,
		minlen=70
	threads: 8
	priority: 60
	shell:
		"""
		java -Xmx16g -jar {ALB_TRIM_DIR} PE -threads {threads} -phred33 -trimlog {log} {input.fwd} {input.rev} \
		{output.fwd_pa} {output.fwd_un} {output.rev_pa} {output.rev_un} ILLUMINACLIP:{ADAPT} LEADING:{params.leading} TRAILING:{params.trailing} \
		SLIDINGWINDOW:{WINDOW} MINLEN:{params.minlen}
		"""		


rule bwa:
	input:
		ref=REF,
		fwd=rules.trimmomatic.output.fwd_pa,
		rev=rules.trimmomatic.output.rev_pa
	output:
		temp("mapped/{S}.bam")
	message: """--- Mapping {input.fwd} {input.rev} onto {input.ref} with BWA MEM."""
	params:
		rg="@RG\tID:{S}\tSM:{S}"
	priority: 70
	threads: 10
	shell:
		"""
		{ALB_BWA_DIR} mem -M -R '{params.rg}' -t {threads} {input.ref} {input.fwd} {input.rev} | {ALB_SAMT_DIR} view -Sb - > {output}
		"""


rule flagstat:
	input:
		rules.bwa.output
	output:
		"stats/{S}.flagstat.txt"
	message:
		"""--- Calculating metrics of {input} with Samtools flagstat."""
	shell:
		"{ALB_SAMT_DIR} flagstat {input} > {output}"
		
		
rule stats:
	input:
		rules.bwa.output
	output:
		"stats/{S}.stats.txt"
	message:
		"""--- Calculating metrics of {input} with Samtools flagstat."""
	shell:
		"{ALB_SAMT_DIR} flagstat {input} > {output}"


rule mq20:
	input:
		rules.bwa.output
	output:
		temp("sorted/{S}_sorted20.bam")
	message:
		"""--- Applying MQ20 filter and sorting {input} with Samtools"""
	priority: 80
	threads: 6
	shell:
		"{ALB_SAMT_DIR} view -u -h -F 0x04 -q 20 {input} | {ALB_SAMT_DIR} sort -@ {threads} - sorted/{wildcards.S}_sorted20"


rule targets:
	input: 
		gen="genome/genome_contigs.txt",
		cov="genome/cov_blast_conts.txt",
		ref=REF
	output: 
		tar="genome/tmp_target.fasta",
		ntar="genome/NONtarget.fasta",
	message: "--- Generating fastas for target and nontarget contigs with SEQTK."
	priority: 90
	threads: 2
	shell:
		"""
		cut -f 1 {input.gen} | grep -Fwf {input.cov} - | {ALB_STK_DIR} subseq {input.ref} - > {output.tar}
		cut -f 1 {input.gen} | grep -v -Fwf {input.cov} - | {ALB_STK_DIR} subseq {input.ref} - > {output.ntar}
		"""


rule merge_fasta:
	input:	rules.targets.output.ntar
	output: "genome/NONtarget.fa"
	message:
		"""--- Merging non-target contigs of {input}."""
	priority: 90
	threads: 4
	shell:
		"""
		/usr/local/packages/anaconda2/bin/python scripts/merge_superfasta.py
		/usr/local/packages/anaconda2/bin/python scripts/edit_superfasta.py
		"""


rule cat_fasta:
	input:
		tar=rules.targets.output.tar,
		ntar=rules.merge_fasta.output
	output:
		"red_ref/NEW_superscaffold_REF.fasta"
	message:
		"""--- Concatenating target and non-target contigs into {output}."""
	priority: 90
	threads: 2
	shell:
		"""
		cat {input.tar} {input.ntar} > {output}
		"""



rule red_prep:
	input: rules.cat_fasta.output
	output: "red_ref/NEW_superscaffold_REF.fasta.fai"
	message:
		"""--- Preparation {input} with BWA index, samtools faidx, and Picard dict."""
	priority: 90
	threads: 5
	shell:
		"""
		/bin/sh scripts/red_prep.sh {input} {output}
		"""


rule int_list:
	input: rules.red_prep.output
	output: "Target_contigs.bed"
	message: """--- Generating {output} from {input}."""
	priority: 90
	shell:
		"""
		/bin/sh scripts/int_list.sh {input} {output}
		"""


rule var_call:
	input: 
		bam="clip2red/{S}_clip.bam",
		ref="red_ref/NEW_superscaffold_REF.fasta",
		tar="Target_contigs.bed"
	output: "results/gVCF/{S}.g.vcf.gz"
	message: """--- Calling SNPs and indels in {input} with HaplotypeCaller"""
	priority: 70
	threads: 5
	shell:
		"""
		java -Xmx34g -jar {ALB_GATK_DIR} HaplotypeCaller -R {input.ref} -I {input.bam} \
		-O {output} -ERC GVCF --heterozygosity 0.05 --pcr-indel-model NONE -L {input.tar}
		"""


def findGVCF ():
	gvcfs = open("samplesGVCF.list", "w")
	gvcfList = glob.glob(RES_DIR + '/*')
	for sample in gvcfList:
		if sample.endswith(".g.vcf.gz"):
			print>>gvcfs, sample
	gvcfs.close()
	return ["samplesGVCF.list"];


rule CombineGVCFs:
	input:
		#gvcf=findGVCF,
		gvcf="samplesGVCF_all.list",
		ref="red_ref/NEW_superscaffold_REF.fasta"
	output:
		"results/CZA_all.g.vcf.gz"
	message: """--- Combine per-sample gVCFs into a multi-sample gVCF file with CombineGVCFs."""
	priority: 70
	threads: 7
	shell:
		"""
		java -Xmx34g -jar {ALB_GATK_DIR} CombineGVCFs -R {input.ref} -V {input.gvcf} -O {output}
		"""



#rule CombineGVCFall:
#	input:
#		#gvcf=findGVCF,
#		gvcf1="results/CZA.g.vcf.gz",
#		gvcf2="results/CZA_02.g.vcf.gz",
#		ref="red_ref/NEW_superscaffold_REF.fasta"
#	output:
#		"results/CZA_all.g.vcf.gz"
#	message: """--- Combine combined gVCFs into the final multi-sample gVCF file with CombineGVCFs."""
#	priority: 70
#	threads: 7
#	shell:
#		"""
#		java -Xmx34g -jar {ALB_GATK_DIR} CombineGVCFs -R {input.ref} -V {input.gvcf1} -V {input.gvcf2} -O {output}
#		"""


rule GenotypeGVCFs:
	input:
		ref="red_ref/NEW_superscaffold_REF.fasta",
		gvcf=rules.CombineGVCFs.output,
		tar="Target_contigs.bed"
	output:
		"results/VCF/CZA_raw_all.vcf.gz"
	message: """--- Perform joint genotyping on all samples with GenotypeGVCFs."""
	priority: 70
	threads: 7
	shell:
		"""
		java -Xmx34g -jar {ALB_GATK_DIR} GenotypeGVCFs -R {input.ref} -V {input.gvcf} -O {output} \
		--heterozygosity 0.05 --only-output-calls-starting-in-intervals TRUE -L {input.tar}
		"""
